platform/machine/stackit/cluster.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/stackitcloud/stackit-sdk-go/services/iaas"
+	"k8s.io/utils/ptr"
+)
+
+type cluster struct {
+	*platform.BaseCluster
+	flight  *flight
+	network *stackit.Network
+	keypair *stackit.Keypair
+}
+
+func (bc *cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error) {
+	ctx := context.TODO()
+
+	userDataConf, err := bc.RenderUserData(userdata, map[string]string{
+		"$public_ipv4":  "${COREOS_CUSTOM_PUBLIC_IPV4}",
+		"$private_ipv4": "${COREOS_CUSTOM_PRIVATE_IPV4}",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	userDataConf.AddSystemdUnitDropin("coreos-metadata.service", "00-custom-metadata.conf", `[Service]
+	Environment=OUTPUT=/run/metadata/flatcar
+	ExecStartPost=/usr/bin/sed -i "s/STACKIT/CUSTOM/" /run/metadata/flatcar
+	ExecStartPost=/usr/bin/sed -i "s/PRIVATE_IPV4_0/PRIVATE_IPV4/" /run/metadata/flatcar
+	ExecStartPost=/usr/bin/sed -i "s/PUBLIC_IPV4_0/PUBLIC_IPV4/" /run/metadata/flatcar
+	ExecStartPost=/usr/bin/sed -i "s#/32##" /run/metadata/flatcar
+	ExecStart=/usr/bin/bash -c 'echo "COREOS_CUSTOM_PRIVATE_IPV4=$(ip addr show eth0 | grep "brd" | grep -Po "inet \K[\d.]+")" > ${OUTPUT}'
+	ExecStartPost=/usr/bin/sh -c "ip addr add $(cat /run/metadata/flatcar | grep COREOS_CUSTOM_PRIVATE_IPV4 | cut -d '=' -f 2) dev eth0"
+	`)
+
+	base64Config := make([]byte, base64.StdEncoding.EncodedLen(len(userDataConf.Bytes())))
+	base64.StdEncoding.Encode(base64Config, userDataConf.Bytes())
+
+	secGroup, err := bc.flight.api.CreateSecurityGroup(ctx, "flatcar_security_group")
+	if err != nil {
+		return nil, fmt.Errorf("error creating security group: %s", err)
+	}
+	err = bc.flight.api.CreateSecurityGroupRule(ctx, *secGroup.Id)
+	if err != nil {
+		return nil, fmt.Errorf("error creating security group rule: %s", err)
+	}
+
+	ipAddress, err := bc.flight.api.CreateIPAddress(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error creating IP address: %s", err)
+	}
+
+	var keyPairName iaas.CreateServerPayloadGetKeypairNameAttributeType
+	if bc.keypair != nil {
+		keyPairName = bc.keypair.Name
+	}
+	instance, err := bc.flight.api.CreateServer(ctx, bc.vmname(), bc.network.NetworkId, keyPairName, &base64Config)
+	if err != nil {
+		return nil, err
+	}
+
+	err = bc.flight.api.AddSecurityGroup(ctx, *instance.Id, *secGroup.Id)
+	if err != nil {
+		return nil, fmt.Errorf("error adding security group: %s", err)
+	}
+
+	err = bc.flight.api.AttachPublicIPAddress(ctx, *ipAddress.Id, *instance.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	// The API does sometimes need a couple of seconds to report the attached IP address
+	for {
+		instance, err = bc.flight.api.GetServer(ctx, *instance.Id)
+		if err != nil {
+			return nil, fmt.Errorf("error getting server: %s", err)
+		}
+		if !instance.HasNics() {
+			return nil, fmt.Errorf("no NICs available")
+		}
+		hasPublicIP := false
+		for _, nic := range instance.GetNics() {
+			if nic.HasPublicIp() {
+				hasPublicIP = true
+			}
+		}
+		if hasPublicIP {
+			break
+		}
+	}
```

Comment:
This can potentially block the whole execution (if for whatever reasons the API does not report back the public IP and/or the NIC).
We can use this helper here:
```
$ go doc github.com/flatcar/mantle/util.Retry
package util // import "github.com/flatcar/mantle/util"

func Retry(attempts int, delay time.Duration, f func() error) error
    Retry calls function f until it has been called attemps times, or succeeds.
    Retry delays for delay between calls of f. If f does not succeed after
    attempts calls, the error from the last call is returned.
```